File: scripts/rakuten-rms-extract.ts
/**
 * Rakuten RMS Daily Sales & Traffic Extraction
 *
 * Required env vars:
 *   RAKUTEN_SERVICE_SECRET  - RMS サービスシークレット
 *   RAKUTEN_LICENSE_KEY     - RMS ライセンスキー
 *
 * Optional:
 *   TARGET_DATE  - YYYYMMDD (defaults to yesterday)
 *
 * Usage:
 *   npx tsx scripts/rakuten-rms-extract.ts
 */

import * as fs from 'fs';
import * as path from 'path';

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

const SERVICE_SECRET = process.env.RAKUTEN_SERVICE_SECRET ?? '';
const LICENSE_KEY    = process.env.RAKUTEN_LICENSE_KEY ?? '';

const TARGET_DATE: string = (() => {
  if (process.env.TARGET_DATE) return process.env.TARGET_DATE;
  const d = new Date();
  d.setDate(d.getDate() - 1);
  return d.toISOString().slice(0, 10).replace(/-/g, '');
})();

// YYYYMMDD → YYYY-MM-DD HH:MM:SS (start/end of day JST = UTC+9, stored as UTC)
const dateToRangeJST = (yyyymmdd: string) => {
  const y = yyyymmdd.slice(0, 4);
  const m = yyyymmdd.slice(4, 6);
  const d = yyyymmdd.slice(6, 8);
  // JST 00:00 = UTC 2026-06-26 15:00, JST 23:59:59 = UTC 2026-06-27 14:59:59
  return {
    from: `${y}-${m}-${d}T00:00:00+09:00`,
    to:   `${y}-${m}-${d}T23:59:59+09:00`,
  };
};

const OUTPUT_DIR = path.join(process.cwd(), 'data', 'rakuten');
const RMS_BASE   = 'https://api.rms.rakuten.co.jp/es/2.0';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

interface OrderItem {
  itemUrl: string;
  itemName: string;
  units:    number;
  price:    number;
}

interface Order {
  orderNumber:  string;
  orderDatetime: string;
  orderItems:   OrderItem[];
}

interface SalesSummaryRow {
  item_url:      string;
  item_name:     string;
  units_sold:    number;
  gross_sales:   number;
}

interface AccessRow {
  item_url:     string;
  access_count: number;
  note:         string;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

async function rmsGet(path: string, authHeader: string): Promise<unknown> {
  const url = `${RMS_BASE}${path}`;
  const res = await fetch(url, {
    headers: {
      Authorization: authHeader,
      'Content-Type': 'application/json; charset=utf-8',
    },
  });
  if (!res.ok) {
    const body = await res.text();
    throw new Error(`RMS API error ${res.status} for ${url}: ${body}`);
  }
  return res.json();
}

async function rmsPost(path: string, authHeader: string, body: unknown): Promise<unknown> {
  const url = `${RMS_BASE}${path}`;
  const res = await fetch(url, {
    method: 'POST',
    headers: {
      Authorization: authHeader,
      'Content-Type': 'application/json; charset=utf-8',
    },
    body: JSON.stringify(body),
  });
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`RMS API error ${res.status} for POST ${url}: ${text}`);
  }
  return res.json();
}

function csvEscape(v: string | number): string {
  const s = String(v);
  return s.includes(',') || s.includes('"') || s.includes('\n')
    ? `"${s.replace(/"/g, '""')}"`
    : s;
}

function writeCsv(filePath: string, headers: string[], rows: (string | number)[][]): void {
  const bom = '﻿';
  const lines = [
    headers.map(csvEscape).join(','),
    ...rows.map(r => r.map(csvEscape).join(',')),
  ];
  fs.writeFileSync(filePath, bom + lines.join('\n'), 'utf-8');
}

// ---------------------------------------------------------------------------
// Step 1: Fetch orders for the target day (sales by product)
// ---------------------------------------------------------------------------

async function fetchSalesByProduct(authHeader: string): Promise<SalesSummaryRow[]> {
  const { from, to } = dateToRangeJST(TARGET_DATE);

  // RMS Order Search API (searchOrder)
  const payload = {
    orderProgressList:    [100, 200, 300, 400, 500, 600, 700, 800], // all statuses
    dateType:             1, // order date
    startDatetime:        from,
    endDatetime:          to,
    PaginationRequestModel: { requestRecordsAmount: 1000, requestPage: 1 },
  };

  const data = any = await rmsPost('/order/searchOrder/', authHeader, payload);

  const orderNumberList: string[] = data?.orderNumberList ?? [];
  if (orderNumberList.length === 0) {
    console.log('  No orders found for the target date.');
    return [];
  }

  console.log(`  Found ${orderNumberList.length} order(s). Fetching details…`);

  // Batch-fetch order details (up to 100 at a time)
  const orders: Order[] = [];
  const chunkSize = 100;
  for (let i = 0; i < orderNumberList.length; i += chunkSize) {
    const chunk = orderNumberList.slice(i, i + chunkSize);
    const detail: any = await rmsPost('/order/getOrder/', authHeader, { orderNumberList: chunk });
    const orderModelList: any[] = detail?.orderModelList ?? [];
    for (const o of orderModelList) {
      const items: OrderItem[] = (o.PackageModelList ?? []).flatMap((pkg: any) =>
        (pkg.ItemModelList ?? []).map((item: any) => ({
          itemUrl:  item.itemUrl  ?? '',
          itemName: item.itemName ?? '',
          units:    Number(item.units     ?? 0),
          price:    Number(item.priceTaxInc ?? item.price ?? 0) * Number(item.units ?? 0),
        }))
      );
      orders.push({
        orderNumber:   o.orderNumber ?? '',
        orderDatetime: o.orderDatetime ?? '',
        orderItems:    items,
      });
    }
  }

  // Aggregate by item URL
  const map = new Map<string, SalesSummaryRow>();
  for (const order of orders) {
    for (const item of order.orderItems) {
      const key = item.itemUrl || item.itemName;
      const existing = map.get(key);
      if (existing) {
        existing.units_sold  += item.units;
        existing.gross_sales += item.price;
      } else {
        map.set(key, {
          item_url:    item.itemUrl,
          item_name:   item.itemName,
          units_sold:  item.units,
          gross_sales: item.price,
        });
      }
    }
  }

  return Array.from(map.values()).sort((a, b) => b.units_sold - a.units_sold);
}

// ---------------------------------------------------------------------------
// Step 2: Fetch access counts by product (item)
//
// Rakuten does not expose real-time per-item access counts via the standard
// RMS Web API (v2). The access statistics are available in:
//   - RMS > アクセス分析 (web UI only)
//   - Rakuten Analytics API (separate contract, endpoint different)
//
// This function attempts the Item API to list active items, then annotates
// with a note explaining the limitation.
// ---------------------------------------------------------------------------

async function fetchAccessByProduct(authHeader: string): Promise<AccessRow[]> {
  // Attempt: item list endpoint (confirms API connectivity for item data)
  const data: any = await rmsGet('/item/search/?hits=100&offset=0', authHeader);
  const items: any[] = data?.itemSearchResult?.itemList ?? data?.items ?? [];

  if (items.length === 0) {
    console.log('  No items returned from Item API (or endpoint not available).');
    return [];
  }

  console.log(`  Retrieved ${items.length} item(s) from Item API.`);
  console.log('  NOTE: Per-item access counts require Rakuten Analytics API (separate contract).');
  console.log('        Returning item list with placeholder access counts.');

  return items.map((item: any) => ({
    item_url:     item.itemUrl ?? item.manageNumber ?? '',
    access_count: 0, // not available via standard RMS Web API
    note:         'access count not available via RMS Web API v2; use Rakuten Analytics API',
  }));
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function main(): Promise<void> {
  const nowIso = new Date().toISOString();
  console.log('='.repeat(60));
  console.log('Rakuten RMS Daily Data Extraction');
  console.log(`Run timestamp : ${nowIso}`);
  console.log(`Target date   : ${TARGET_DATE} (JST)`);
  console.log('='.repeat(60));

  fs.mkdirSync(OUTPUT_DIR, { recursive: true });

  // -------------------------------------------------------------------------
  // Guard: credentials required
  // -------------------------------------------------------------------------
  if (!SERVICE_SECRET || !LICENSE_KEY) {
    const msg = [
      `Rakuten RMS Extraction Attempt`,
      `Timestamp   : ${nowIso}`,
      `Target date : ${TARGET_DATE}`,
      `Status      : FAILED`,
      `Reason      : Missing credentials.`,
      ``,
      `Required environment variables (not set):`,
      `  RAKUTEN_SERVICE_SECRET  - RMS > API設定 > サービスシークレット`,
      `  RAKUTEN_LICENSE_KEY     - RMS > API設定 > ライセンスキー`,
      ``,
      `How to set them:`,
      `  1. Log into RMS (https://rms.rakuten.co.jp)`,
      `  2. Go to [店舗設定] > [API設定]`,
      `  3. Copy serviceSecret and licenseKey`,
      `  4. Set env vars and re-run:`,
      `     RAKUTEN_SERVICE_SECRET=xxx RAKUTEN_LICENSE_KEY=yyy npx tsx scripts/rakuten-rms-extract.ts`,
    ].join('\n');

    const logFile = path.join(OUTPUT_DIR, `extraction-attempt-${TARGET_DATE}.txt`);
    fs.writeFileSync(logFile, msg, 'utf-8');

    console.error('\n[FAILED] Missing Rakuten RMS credentials.\n');
    console.error(msg);
    console.log(`\nAttempt logged to: ${logFile}`);
    process.exit(1);
  }

  const authHeader = `ESA ${Buffer.from(`${SERVICE_SECRET}:${LICENSE_KEY}`).toString('base64')}`;

  // -------------------------------------------------------------------------
  // Step 1: Sales by product
  // -------------------------------------------------------------------------
  console.log('\n[1/3] Fetching sales (orders) by product…');
  let salesRows: SalesSummaryRow[] = [];
  let salesError = '';
  try {
    salesRows = await fetchSalesByProduct(authHeader);
    console.log(`  → ${salesRows.length} product(s) with sales.`);
  } catch (err: any) {
    salesError = String(err.message ?? err);
    console.error(`  ERROR: ${salesError}`);
  }

  // -------------------------------------------------------------------------
  // Step 2: Access count by product
  // -------------------------------------------------------------------------
  console.log('\n[2/3] Fetching access counts by product…');
  let accessRows: AccessRow[] = [];
  let accessError = '';
  try {
    accessRows = await fetchAccessByProduct(authHeader);
    console.log(`  → ${accessRows.length} product(s) listed.`);
  } catch (err: any) {
    accessError = String(err.message ?? err);
    console.error(`  ERROR: ${accessError}`);
  }

  // -------------------------------------------------------------------------
  // Step 3: Export CSV files
  // -------------------------------------------------------------------------
  console.log('\n[3/3] Exporting CSV files…');

  const salesCsvPath = path.join(OUTPUT_DIR, `sales-by-product-${TARGET_DATE}.csv`);
  if (salesRows.length > 0) {
    writeCsv(salesCsvPath, ['item_url', 'item_name', 'units_sold', 'gross_sales_jpy'], [
      ...salesRows.map(r => [r.item_url, r.item_name, r.units_sold, r.gross_sales]),
    ]);
    console.log(`  Sales CSV  : ${salesCsvPath}`);
  } else if (!salesError) {
    // Write empty CSV with headers
    writeCsv(salesCsvPath, ['item_url', 'item_name', 'units_sold', 'gross_sales_jpy'], []);
    console.log(`  Sales CSV  : ${salesCsvPath} (no sales data)`);
  }

  const accessCsvPath = path.join(OUTPUT_DIR, `access-by-product-${TARGET_DATE}.csv`);
  if (accessRows.length > 0) {
    writeCsv(accessCsvPath, ['item_url', 'access_count', 'note'], [
      ...accessRows.map(r => [r.item_url, r.access_count, r.note]),
    ]);
    console.log(`  Access CSV : ${accessCsvPath}`);
  } else if (!accessError) {
    writeCsv(accessCsvPath, ['item_url', 'access_count', 'note'], []);
    console.log(`  Access CSV : ${accessCsvPath} (no access data)`);
  }

  // Summary report
  const summaryPath = path.join(OUTPUT_DIR, `summary-${TARGET_DATE}.txt`);
  const totalUnits  = salesRows.reduce((s, r) => s + r.units_sold, 0);
  const totalSales  = salesRows.reduce((s, r) => s + r.gross_sales, 0);
  const summary = [
    `Rakuten RMS Daily Extraction Summary`,
    `====================================`,
    `Timestamp   : ${nowIso}`,
    `Target date : ${TARGET_DATE} (JST)`,
    ``,
    `[Sales]`,
    salesError
      ? `  Status     : ERROR - ${salesError}`
      : `  Status     : OK`,
    `  Products   : ${salesRows.length}`,
    `  Total units: ${totalUnits}`,
    `  Total sales: ¥${totalSales.toLocaleString()}`,
    `  CSV file   : ${salesCsvPath}`,
    ``,
    `[Access Counts]`,
    accessError
      ? `  Status     : ERROR - ${accessError}`
      : `  Status     : OK (see note — per-item access requires Rakuten Analytics API)`,
    `  Items      : ${accessRows.length}`,
    `  CSV file   : ${accessCsvPath}`,
    ``,
    `Top 5 products by units sold:`,
    ...salesRows.slice(0, 5).map((r, i) =>
      `  ${i + 1}. ${r.item_name || r.item_url} — ${r.units_sold} units / ¥${r.gross_sales.toLocaleString()}`
    ),
  ].join('\n');

  fs.writeFileSync(summaryPath, summary, 'utf-8');
  console.log(`  Summary    : ${summaryPath}`);

  console.log('\n' + summary);
  console.log('\n[DONE]');
}

main().catch(err => {
  console.error('\n[FATAL]', err);
  process.exit(1);
});

File: scripts/rakuten_rms_extract.py
#!/usr/bin/env python3
"""
Rakuten RMS daily sales & traffic extractor.

Pulls yesterday's order data (sales by product) from the Rakuten RMS Order
Search API and optionally merges access analytics from a CSV downloaded via
RMS > アクセス分析 > 商品別アクセス分析.

Required environment variables:
  RAKUTEN_SERVICE_SECRET   – RMS APIサービスシークレット
  RAKUTEN_LICENSE_KEY      – RMS APIライセンスキー

Optional:
  RAKUTEN_ACCESS_CSV       – Path to access analytics CSV exported from RMS
                             (if omitted, access columns are left blank)
  RAKUTEN_OUTPUT_DIR       – Directory for output CSV (default: ./data/rakuten)
  RAKUTEN_TARGET_DATE      – Override target date as YYYY-MM-DD (default: yesterday)

Usage:
  python3 scripts/rakuten_rms_extract.py
"""

import base64
import csv
import json
import os
import sys
from datetime import date, datetime, timedelta, timezone
from pathlib import Path

try:
    import requests
except ImportError:
    sys.exit("requests is required: pip install requests")

# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------

JST = timezone(timedelta(hours=9))
RMS_API_BASE = "https://api.rms.rakuten.co.jp"
ORDER_SEARCH_ENDPOINT = f"{RMS_API_BASE}/es/2.0/order/searchOrder/"

SERVICE_SECRET = os.environ.get("RAKUTEN_SERVICE_SECRET", "")
LICENSE_KEY = os.environ.get("RAKUTEN_LICENSE_KEY", "")
ACCESS_CSV_PATH = os.environ.get("RAKUTEN_ACCESS_CSV", "")
OUTPUT_DIR = Path(os.environ.get("RAKUTEN_OUTPUT_DIR", "data/rakuten"))

_target_env = os.environ.get("RAKUTEN_TARGET_DATE", "")
if _target_env:
    TARGET_DATE = date.fromisoformat(_target_env)
else:
    TARGET_DATE = date.today() - timedelta(days=1)

# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _auth_header() -> str:
    token = base64.b64encode(f"{SERVICE_SECRET}:{LICENSE_KEY}".encode()).decode()
    return f"ESA {token}"


def _jst_range(d: date) -> tuple[str, str]:
    """Return (start, end) ISO-8601 strings in JST for an entire calendar day."""
    start = datetime(d.year, d.month, d.day, 0, 0, 0, tzinfo=JST)
    end = datetime(d.year, d.month, d.day, 23, 59, 59, tzinfo=JST)
    fmt = "%Y-%m-%dT%H:%M:%S+09:00"
    return start.strftime(fmt), end.strftime(fmt)


# ---------------------------------------------------------------------------
# Step 1 – Fetch orders from RMS Order Search API
# ---------------------------------------------------------------------------


def fetch_orders(target: date) -> list[dict]:
    """
    Return a flat list of order-item dicts for all orders placed on *target*.
    Paginates automatically (1 000 records per page).
    """
    if not SERVICE_SECRET or not LICENSE_KEY:
        print("[WARN] RAKUTEN_SERVICE_SECRET / RAKUTEN_LICENSE_KEY not set.")
        print("       Skipping API call – sales data will be empty.")
        return []

    start_dt, end_dt = _jst_range(target)
    headers = {
        "Authorization": _auth_header(),
        "Content-Type": "application/json; charset=utf-8",
    }

    all_items: list[dict] = []
    page = 1

    while True:
        payload = {
            "dateType": 1,            # 1 = 注文日
            "startDatetime": start_dt,
            "endDatetime": end_dt,
            "PaginationRequestModel": {
                "requestRecordsAmount": 1000,
                "requestPage": page,
            },
        }

        try:
            resp = requests.post(
                ORDER_SEARCH_ENDPOINT,
                headers=headers,
                json=payload,
                timeout=30,
            )
        except requests.RequestException as exc:
            print(f"[ERROR] RMS API request failed: {exc}")
            break

        if resp.status_code != 200:
            print(f"[ERROR] RMS API returned {resp.status_code}: {resp.text[:400]}")
            break

        body = resp.json()

        # Surface any RMS-level error codes
        message_model = body.get("MessageModel", {})
        if message_model.get("messageType") == "ERROR":
            print(f"[ERROR] RMS API error: {message_model}")
            break

        order_list = body.get("orderNumberList", []) or []
        # The v2 searchOrder returns orderNumberList; actual details need
        # /order/getOrder/ but some responses embed OrderModelList directly.
        order_models = body.get("OrderModelList", []) or []

        for order in order_models:
            order_number = order.get("orderNumber", "")
            order_date = order.get("orderDatetime", "")[:10]
            for pkg in order.get("packageModelList", []) or []:
                for item in pkg.get("itemModelList", []) or []:
                    all_items.append(
                        {
                            "order_number": order_number,
                            "order_date": order_date,
                            "item_id": item.get("itemId", ""),
                            "item_number": item.get("itemNumber", ""),
                            "item_name": item.get("itemName", ""),
                            "units": int(item.get("units", 0)),
                            "unit_price": int(item.get("unitPrice", 0)),
                            "subtotal": int(item.get("price", 0)),
                        }
                    )

        # Check if there are more pages
        pagination = body.get("PaginationResponseModel", {})
        total_pages = pagination.get("totalPages", 1)
        if page >= total_pages:
            break
        page += 1

    return all_items


# ---------------------------------------------------------------------------
# Step 2 – Aggregate sales by product
# ---------------------------------------------------------------------------


def aggregate_sales(order_items: list[dict]) -> dict[str, dict]:
    """
    Roll up order items to a per-product summary keyed by item_number.
    """
    summary: dict[str, dict] = {}

    for item in order_items:
        key = item["item_number"] or item["item_id"]
        if key not in summary:
            summary[key] = {
                "item_number": key,
                "item_name": item["item_name"],
                "order_count": 0,
                "units_sold": 0,
                "total_sales": 0,
            }
        summary[key]["order_count"] += 1
        summary[key]["units_sold"] += item["units"]
        summary[key]["total_sales"] += item["subtotal"]

    return summary


# ---------------------------------------------------------------------------
# Step 3 – Load access analytics (optional CSV from RMS UI)
# ---------------------------------------------------------------------------
#
# Access analytics are NOT available through the standard RMS REST API.
# To obtain them:
#   RMS > ショップマネージャー > アクセス分析 > 商品別アクセス分析
#   → 日別 → 対象日を選択 → CSVダウンロード
#
# Expected columns (Rakuten standard export):
#   商品番号, 商品名, PV数, ユニークユーザー数, カート投入数
#


def load_access_csv(csv_path: str) -> dict[str, dict]:
    """
    Parse the access analytics CSV from RMS and return a dict keyed by
    item_number (商品番号).
    """
    if not csv_path:
        return {}

    path = Path(csv_path)
    if not path.exists():
        print(f"[WARN] Access CSV not found: {csv_path}")
        return {}

    access: dict[str, dict] = {}
    try:
        with open(path, encoding="utf-8-sig") as f:
            reader = csv.DictReader(f)
            for row in reader:
                # Try common Rakuten column names
                item_no = (
                    row.get("商品番号")
                    or row.get("item_number")
                    or row.get("管理番号")
                    or ""
                ).strip()
                if not item_no:
                    continue
                access[item_no] = {
                    "page_views": _int(row, ["PV数", "page_views", "PV"]),
                    "unique_users": _int(row, ["ユニークユーザー数", "unique_users", "UU数"]),
                    "cart_adds": _int(row, ["カート投入数", "cart_adds"]),
                }
    except Exception as exc:
        print(f"[WARN] Failed to parse access CSV: {exc}")

    return access


def _int(row: dict, keys: list[str]) -> int:
    for k in keys:
        v = row.get(k, "").replace(",", "").strip()
        if v.isdigit():
            return int(v)
    return 0


# ---------------------------------------------------------------------------
# Step 4 – Merge and export
# ---------------------------------------------------------------------------


def export_csv(
    sales: dict[str, dict],
    access: dict[str, dict],
    target: date,
    output_dir: Path,
) -> Path:
    output_dir.mkdir(parents=True, exist_ok=True)
    filename = output_dir / f"rakuten_daily_{target.strftime('%Y%m%d')}.csv"

    all_keys = sorted(set(sales) | set(access))

    fieldnames = [
        "date",
        "item_number",
        "item_name",
        "units_sold",
        "order_count",
        "total_sales",
        "page_views",
        "unique_users",
        "cart_adds",
    ]

    with open(filename, "w", newline="", encoding="utf-8-sig") as f:
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()

        for key in all_keys:
            s = sales.get(key, {})
            a = access.get(key, {})
            writer.writerow(
                {
                    "date": target.isoformat(),
                    "item_number": key,
                    "item_name": s.get("item_name", a.get("item_name", "")),
                    "units_sold": s.get("units_sold", 0),
                    "order_count": s.get("order_count", 0),
                    "total_sales": s.get("total_sales", 0),
                    "page_views": a.get("page_views", ""),
                    "unique_users": a.get("unique_users", ""),
                    "cart_adds": a.get("cart_adds", ""),
                }
            )

    return filename


# ---------------------------------------------------------------------------
# Step 5 – Console summary table
# ---------------------------------------------------------------------------


def print_summary(
    sales: dict[str, dict],
    access: dict[str, dict],
    target: date,
    output_path: Path,
) -> None:
    all_keys = sorted(set(sales) | set(access))

    print()
    print(f"=== Rakuten RMS Daily Report  {target.isoformat()} ===")
    print()

    if not all_keys:
        print("  No data found for this date.")
        print("  Possible reasons:")
        print("  - No orders were placed on this date")
        print("  - API credentials are missing or incorrect")
        print("  - RAKUTEN_ACCESS_CSV was not provided")
    else:
        col = "{:<20} {:<40} {:>8} {:>7} {:>12} {:>8} {:>8} {:>8}"
        header = col.format(
            "Item Number", "Item Name", "Units", "Orders",
            "Sales (¥)", "PV", "UU", "Cart"
        )
        print(header)
        print("-" * len(header))

        total_units = total_orders = total_sales = 0
        for key in all_keys:
            s = sales.get(key, {})
            a = access.get(key, {})
            name = (s.get("item_name") or a.get("item_name") or "")[:38]
            u = s.get("units_sold", 0)
            o = s.get("order_count", 0)
            rev = s.get("total_sales", 0)
            pv = a.get("page_views", "-")
            uu = a.get("unique_users", "-")
            cart = a.get("cart_adds", "-")
            total_units += u
            total_orders += o
            total_sales += rev
            print(col.format(key[:20], name, u, o, f"{rev:,}", pv, uu, cart))

        print("-" * len(header))
        print(col.format("TOTAL", "", total_units, total_orders, f"{total_sales:,}", "", "", ""))

    print()
    print(f"Output saved → {output_path.resolve()}")
    print()


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------


def main() -> None:
    print(f"[INFO] Target date : {TARGET_DATE.isoformat()}")
    print(f"[INFO] Output dir  : {OUTPUT_DIR.resolve()}")
    print()

    # Step 1 & 2 – orders
    print("[1/3] Fetching orders from Rakuten RMS Order API …")
    order_items = fetch_orders(TARGET_DATE)
    print(f"      → {len(order_items)} line-item(s) retrieved")

    sales = aggregate_sales(order_items)
    print(f"      → {len(sales)} unique product(s) sold")

    # Step 3 – access data
    print("[2/3] Loading access analytics …")
    if ACCESS_CSV_PATH:
        access = load_access_csv(ACCESS_CSV_PATH)
        print(f"      → {len(access)} product(s) with access data")
    else:
        access = {}
        print("      → RAKUTEN_ACCESS_CSV not set; skipping access data")
        print("        (Download from RMS > アクセス分析 > 商品別アクセス分析 > CSVダウンロード)")

    # Step 4 – export
    print("[3/3] Writing CSV …")
    out = export_csv(sales, access, TARGET_DATE, OUTPUT_DIR)
    print(f"      → {out.resolve()}")

    # Summary
    print_summary(sales, access, TARGET_DATE, out)


if __name__ == "__main__":
    main()
